File: DSA/Python/LinkedList/reverse_linked_list.py
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next
        
# Logic in solution is simple we simply turn away nodes, and store previous node that we turned away
# on next iteration we link this prev node as next for new current
# in between iteration there is broken connection but we keep reference for an processeced one
def reverseList(head:Optional[ListNode]) -> Optional[ListNode]:
    
    prev = None
    current = head
    
    while current != None:
        # next is 2 | ... | next is 6 | next is None
        next = current.next
        # 1 next none | 5 next is prev(4) | 6 next is prev(5)
        current.next = prev
        # 1 is processed and become prev(with next == None)| prev = 5 | prev = 6
        prev = current
        # now we looking at 2 which is new current| current = 6 | current = None
        current = next
    
    return prev
        

File: DSA/Dotnet/LeetCode.Net/Problems/LinkedList/ReverseLinkedList.cs
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public int val;
 *     public ListNode next;
 *     public ListNode(int val=0, ListNode next=null) {
 *         this.val = val;
 *         this.next = next;
 *     }
 * }
 */
namespace LeetCode.Problems.LinkedList;

public class ReverseLinkedList
{
    public ListNode ReverseList(ListNode head)
    {
        if (head == null)
        {
            return null;
        }

        ListNode prev = null;
        var current = head;

        while (current != null)
        {
            ListNode next = current.next;
            current.next = prev;
            prev = current;
            current = next;
        }

        return prev;
    }
}